numpy/core/src/multiarray/arraytypes.c.src

```diff
@@ -426,16 +440,25 @@ UNICODE_setitem(PyObject *op, void *ov, void *vap)
     char *buffer;
 #endif
 
-    if (!PyBytes_Check(op) && !PyUnicode_Check(op) &&
-            PySequence_Check(op) && PySequence_Size(op) > 0) {
+    /* Handle case of assigning from an array scalar */
+    if (PyArray_Check(op) && PyArray_NDIM((PyArrayObject *)op) == 0) {
```

Comment:
no clue, apparently both worth. Might be worth testing using getitem in the other 3 cases and see if something fails.
As it already are two approaches in the file imo it is enough to move the copied one into a function and add a TODO